File: python-algorithm/leetcode/problem_2.py
"""2. Add Two Numbers
https://leetcode.com/problems/add-two-numbers/

You are given two non-empty linked lists representing two non-negative integers.
The digits are stored in reverse order and each of their nodes contain a single
digit. Add the two numbers and return it as a linked list.

You may assume the two numbers do not contain any leading zero,
except the number 0 itself.

Example:

Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
Output: 7 -> 0 -> 8
Explanation: 342 + 465 = 807.
"""
from typing import Optional

from common.list_node import ListNode


class Solution:
    def add_two_numbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0)
        cur = dummy
        add = 0
        while l1 and l2:
            add, val = divmod(l1.val + l2.val + add, 10)
            cur.next = ListNode(val)
            cur = cur.next
            l1 = l1.next
            l2 = l2.next
        l3 = l1 if l1 else l2
        while l3:
            if add == 0:
                cur.next = l3
                break
            add, val = divmod(l3.val + add, 10)
            cur.next = ListNode(val)
            cur = cur.next
            l3 = l3.next
        if add:
            cur.next = ListNode(add)
        return dummy.next

File: go-algo/leetcode/problem2.go
package main

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	dummy := ListNode{-1, nil}
	p1, p2, p3 := l1, l2, &dummy
	carry := 0
	for p1 != nil && p2 != nil {
		value := p1.Val + p2.Val + carry
		value, carry = value%10, value/10
		node := ListNode{value, nil}
		p3.Next = &node
		p1 = p1.Next
		p2 = p2.Next
		p3 = p3.Next
	}
	if p1 != nil {
		p3.Next = helper(p1, carry)
	} else if p2 != nil {
		p3.Next = helper(p2, carry)
	} else if carry == 1 {
		p3.Next = &ListNode{1, nil}
	}
	return dummy.Next
}

func helper(node *ListNode, carry int) *ListNode {
	if carry == 0 {
		return node
	}
	ptr := node
	for ptr != nil {
		ptr.Val += carry
		if ptr.Val < 10 {
			break
		}
		ptr.Val, carry = ptr.Val%10, ptr.Val/10
		if ptr.Next == nil && carry == 1 {
			ptr.Next = &ListNode{1, nil}
			break
		}
		ptr = ptr.Next
	}
	return node
}
